File: src/server/server-channel.ts
import EventEmitter2 from 'eventemitter2'

import { BifrostEvents, ServerEvents } from '../utils'
import type { ClientNode } from './client-node'
import type { EventOptions } from './event'
import { Event } from './event'
import { getRoomName } from './methods'
import type { Server } from './server'
import { HttpTransportEvents, WebSocketTransportEvents } from './transports'

const SystemEvents: string[] = [
  ...Object.values(HttpTransportEvents),
  ...Object.values(ServerEvents),
  ...Object.values(WebSocketTransportEvents),
]

export class ServerChannel extends EventEmitter2 {
  channelName: string
  server: Server

  constructor(channelName: string) {
    super({
      maxListeners: 1024,
    })

    this.channelName = channelName

    this.onAny((event, value) => {
      if (
        !this.server.events.has(event as string) &&
        !SystemEvents.includes(event as string)
      ) {
        console.warn('Event Not Registered:', event)
      }

      if (this.server.events.has(event as string)) {
        const eventObject = this.server.events.get(event as string)

        eventObject.handler(this, value)
      }
    })
  }

  setServer(server: Server) {
    this.server = server
  }

  /**
   * Broadcasts a pre-encoded event payload to all subscribed clients.
   * The payload already contains the wire protocol `t` field (set by
   * Event.handler), so no re-encoding is needed.
   * @param excludeUuid - Optional client uuid to exclude from receiving the event
   */
  propagate(event: string, payload: string, excludeUuid?: string): void {
    const eventObject = this.server.events.get(event)

    if (!eventObject) {
      console.log('Event Not Registered:', event)
      return
    }

    const transport = this.server.webSocketTransport
    if (!transport) {
      console.warn('[Bifrost] WebSocket transport not available for propagate')
      return
    }

    const roomName = getRoomName(this.channelName, event)

    if (excludeUuid) {
      const excludeClient = this.server.allClients.get(excludeUuid)
      transport.rooms.broadcast(roomName, payload, excludeClient?.socket)
    } else {
      transport.rooms.broadcast(roomName, payload)
    }
  }

  defer<T = any>(event: string, params?: T) {
    process.nextTick(() => {
      this.emit(event, params)
    })
  }

  /** Refreshes a method by its identifier. */
  refresh(method: string) {
    this.emit(BifrostEvents.METHOD_REFRESH, method)
  }

  /** Declares a new event. */
  addEvent(name: string, opts?: EventOptions) {
    if (this.server.events.has(name)) {
      this.server.events.delete(name)
    }

    const event = new Event(name, this.server, this, opts)

    this.server.events.set(name, event)
  }

  get list() {
    return Array.from(this.server.events.keys())
  }

  get length() {
    return this.server.events.size
  }

  get(event: string) {
    return this.server.events.get(event)
  }

  has(event: string) {
    return this.server.events.has(event)
  }

  delete(event: string) {
    return this.server.events.delete(event)
  }

  /**
   * Checks if a client is subscribed to an event via the RoomRegistry.
   */
  isSubscribed(client: ClientNode, event: Event): boolean {
    if (!client.socket) return false

    const rooms = this.server.webSocketTransport?.rooms
    if (!rooms) return false

    const roomName = getRoomName(this.channelName, event.name)
    return rooms.has(client.socket, roomName)
  }
}

File: typeferry-py/src/typeferry/server/server_channel.py
"""Server channel — port of ``typeferry-ts/src/server/server-channel.ts``.

Channels own:

* a registry of events routable on that channel
* :meth:`propagate` — broadcast a pre-encoded event payload to
  subscribed sockets in the corresponding room
* :meth:`emit` — trigger an event's ``handler`` which re-enters this
  channel via ``propagate`` (or Redis, for clustered events)

The TS implementation extends ``EventEmitter2`` with an ``onAny``
interceptor that routes every emit through the registered event's
``handler``. Python doesn't need that pattern — we expose a direct
:meth:`emit` that awaits the handler.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from typeferry.server.event import Event, EventOptions
from typeferry.server.room_name import get_room_name

if TYPE_CHECKING:
    from typeferry.server.client_node import ClientNode
    from typeferry.server.server import Server


class ServerChannel:
    channel_name: str
    server: Server

    def __init__(self, channel_name: str) -> None:
        self.channel_name = channel_name

    def set_server(self, server: Server) -> None:
        self.server = server

    # ------------------------------------------------------------------
    # Emission
    # ------------------------------------------------------------------

    async def emit(self, event: str, params: Any = None) -> None:
        """Trigger the declared event's handler for this channel.

        If the event isn't registered, emit is a no-op with a log line
        (matching the TS behavior).
        """

        event_obj = self.server.events.get(event)
        if event_obj is None:
            return
        await event_obj.handler(self, params)

    async def propagate(
        self, event: str, payload: str, exclude_uuid: str | None = None
    ) -> None:
        """Broadcast a pre-encoded payload to the event's room.

        ``payload`` is the EJSON-encoded ``{t:"event", ...}`` wire frame
        built by :meth:`typeferry.server.event.Event.handler`.
        """

        if self.server.events.get(event) is None:
            return

        transport = self.server.websocket_transport
        if transport is None:
            return

        rooms = transport.rooms
        room_name = get_room_name(self.channel_name, event)

        exclude_socket = None
        if exclude_uuid:
            node = self.server.all_clients.get(exclude_uuid)
            if node is not None:
                exclude_socket = node.socket

        awaitables = rooms.broadcast(room_name, payload, exclude_socket)
        # Individual senders may return coroutines; await the batch so
        # concurrent sends complete before the emit resolves.
        for awaitable in awaitables:
            await awaitable

    # ------------------------------------------------------------------
    # Event registry (delegates to the shared server map)
    # ------------------------------------------------------------------

    def add_event(self, name: str, opts: EventOptions | None = None) -> Event:
        """Declare a new event on this channel. Replaces an existing one."""

        event = Event(name, self.server, self, opts)
        self.server.events[name] = event
        return event

    @property
    def list(self) -> list[str]:
        return list(self.server.events.keys())

    @property
    def length(self) -> int:
        return len(self.server.events)

    def get(self, event: str) -> Event | None:
        return self.server.events.get(event)

    def has(self, event: str) -> bool:
        return event in self.server.events

    def delete(self, event: str) -> bool:
        return self.server.events.pop(event, None) is not None

    def is_subscribed(self, client: ClientNode, event: Event) -> bool:
        if client.socket is None:
            return False
        transport = self.server.websocket_transport
        if transport is None:
            return False
        room_name = get_room_name(self.channel_name, event.name)
        return bool(transport.rooms.has(client.socket, room_name))
